1.2.0 - Add operation handler

janmartenjongerius opened the pull request:
The operation handler is better suited in a service oriented application.
It allows to prepare visitors separate from the invoking code and thus separates
concerns about operations and their visitors.

This can be used instead of a Transaction, directly. By invoking it, it will create a transaction on-the-fly and pass along pre-configured visitors.

```php
<?php
use Johmanx10\Transaction\OperationHandler;
use Johmanx10\Transaction\OperationInterface;
use Johmanx10\Transaction\Visitor\LogOperationVisitor;
use Psr\Log\LoggerInterface;

$handler = new OperationHandler();
$handler->attachVisitor(
    /** @var LoggerInterface $logger */
    new LogOperationVisitor($logger)
);

/** @var OperationInterface[] $operations */
$handler->handle(...$operations);
```

The very goal is that the calling code only needs to know about operations and their handler. Not about visitors. ~[I have some more ideas on how we can clean up the calling code.](https://github.com/johmanx10/transaction/pull/5#issuecomment-600892942) They can be added through a separate pull request.~

It's probably easiest to grasp the pull request by looking at the new example file: https://github.com/johmanx10/transaction/blob/7b760791b4a0435699e16899e6d3f056cb009e31/examples/operation-handler

Closes #3

Commit:
Merge branch 'master' of github.com:johmanx10/transaction

* 'master' of github.com:johmanx10/transaction:
  Make log level configurable
  Add operation visitors to transaction

Commit:
Update GrumPHP configuration

The configuration for GrumPHP has been updated after renewing the
MediaCT testing suite package.

Commit:
Add Operation Handler

A new class, Operation Handler, is introduced to simplify services
focused on executing operations in a transaction.

This is mostly useful when operation visitors can be pre-configured,
allowing the calling service in turn to not be aware of the visitors.
While using a bare Transaction, the visitors are passed on at runtime,
while the operation handler allows to have them attached before the
operations are handled.

Commit:
Add operation handler test

Add a unit test to cover the operation handler.

Commit:
Add operation handler to documentation

Describe how the operation handler works and when to use it instead of a
Transaction directly.

johmanx10 commented:
@ashokadewit I have been thinking about two more ways to simplify this library for the end-user.

1. Create a formatter that accepts any type of exception, but knows to apply a special formatter for certain library specific exceptions. This would simplify the `catch` statements and bodies for the calling code.
2. Make it so the operation handler uses that internally and instead of returning `void`, it will return a result object which exposes whether the transaction succeeded and holds the exception as well as the formatted exception. - Conversely it could throw a more generic `OperationException` which holds the previous exception like any exception does, plus the specifically formatted exception message with the aforementioned formatter.

If either or both of those features seem interesting to you, I will add those to this release as well. My personal preference goes to both *1* and the Exception implementation of *2*.

johmanx10 commented:
It seems this closes #3 , as it replaces the need for a factory in a service oriented application.

janmartenjongerius commented on src/OperationHandler.php:
```diff
@@ -0,0 +1,59 @@
+     */
+    public function handle(OperationInterface ...$operations): void
+    {
+        $transaction = new Transaction(...$operations);
```
I could decouple this by introducing a factory for transactions and optionally injecting that in the constructor of the operation handler. Would you prefer this, @ashokadewit ?

Commit:
Add transaction factory

Add a factory for transactions, so the operation handler does not
determine the concrete transaction class.

Commit:
Upgrade to accepting transaction factory

The transaction factory is moved to the visitor namespace and now
returns an accepting transaction factory.

johmanx10 commented:
After having a look at the available formatters and exceptions, seeing their interfaces aren't able to share a parent interface before PHP 7.4, I have come up with a different approach.

I could inject a factory for creating an operation exception, based on any throwable, into the operation handler. That way, the factory for this exception holds shared logic to format the exception messages, but additionally it can determine the affected operation and pass that along to the operation exception, to further simplify exceptions when debugging failed operations.

I would like the time to make this part of the current pull request.

Commit:
Add operation exception

The operation handler now throws an operation exception, based on the
caught internal exception. This allows the operation exception to be
filled with the formatted version of the previous exception and to
provide the causing operation, when possible.

Commit:
Document and test operation exception

The documentation and example have been updated to reflect the new
behavior of the operation handler, concerning operation exceptions.
